File: src/main/java/com/manifestors/shinrai/client/utils/file/FileManager.java
package com.manifestors.shinrai.client.utils.file;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.manifestors.shinrai.client.Shinrai;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

public class FileManager {

    private static final File ROOT_DIRECTORY = new File("Shinrai");

    public static void createDirectories() {
        if (ROOT_DIRECTORY.mkdirs())
            Shinrai.logger.info("Shinrai directory initialized successfully.");
    }

    public static String toJsonOnlyExposedFields(Object obj) {
        Gson gson = new GsonBuilder()
                .excludeFieldsWithoutExposeAnnotation()
                .setPrettyPrinting()
                .create();

        return gson.toJson(obj);
    }

    @SuppressWarnings("ResultOfMethodCallIgnored")
    public static void writeJsonToFile(String json, String childDirectory, String fileName) {
        try {
            var folder = new File(ROOT_DIRECTORY, childDirectory);
            var jsonFile = new File(folder, fileName);
            folder.mkdirs();
            jsonFile.createNewFile();

            Files.writeString(Path.of(jsonFile.toURI()), json);
        } catch (IOException e) {
            Shinrai.logger.error("Can't write json: ", e);
        }
    }

    public static String getJsonFromFile(String childDirectory, String fileName) {
        try {
            var folder = new File(ROOT_DIRECTORY, childDirectory);
            var jsonFile = new File(folder, fileName);

            return Files.readString(jsonFile.toPath());
        } catch (IOException e) {
            Shinrai.logger.error("Can't read json: ", e);
        }

        return "";
    }
}

File: src/main/kotlin/com/manifestors/shinrai/client/utils/file/FileManager.kt
package com.manifestors.shinrai.client.utils.file

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.JsonElement
import com.manifestors.shinrai.client.Shinrai
import java.io.File
import java.lang.reflect.Type
import java.nio.file.Files
import java.nio.file.Path

object FileManager {
    private val ROOT_DIRECTORY = File("Shinrai")
    private val GSON = Gson()

    fun createDirectories() {
        if (ROOT_DIRECTORY.mkdirs()) Shinrai.logger.info("Shinrai directory initialized successfully.")
    }

    fun toJsonOnlyExposedFields(obj: Any?): String? {
        val gson = GsonBuilder()
            .excludeFieldsWithoutExposeAnnotation()
            .setPrettyPrinting()
            .create()

        return gson.toJson(obj)
    }

    fun <T : Any> toJson(data: T?): String? {
        val gson = GsonBuilder().setPrettyPrinting().create()

        return gson.toJson(data)
    }

    fun writeJsonToFile(json: String, childDirectory: String, fileName: String) {
        runCatching {
            val folder = File(ROOT_DIRECTORY, childDirectory)
            val jsonFile = File(folder, fileName)
            folder.mkdirs()
            jsonFile.createNewFile()

            Files.writeString(Path.of(jsonFile.toURI()), json)
        }.onFailure { e ->
            Shinrai.logger.warn("Can't write json: ", e)
        }
    }

    fun writeJsonToFile(element: JsonElement?, childDirectory: String, fileName: String) {
        runCatching {
            val folder = File(ROOT_DIRECTORY, childDirectory)
            val jsonFile = File(folder, fileName)
            folder.mkdirs()
            jsonFile.createNewFile()

            Files.writeString(Path.of(jsonFile.toURI()), GSON.toJson(element))
        }.onFailure { e ->
            Shinrai.logger.warn("Can't write json: ", e)
        }
    }

    fun getJsonFromFile(childDirectory: String, fileName: String): String {
        runCatching {
            val folder = File(ROOT_DIRECTORY, childDirectory)
            val jsonFile = File(folder, fileName)

            return Files.readString(jsonFile.toPath())
        }.onFailure { e ->
            Shinrai.logger.warn("Can't read json: ", e)
            return ""
        }

        return ""
    }

    fun <T> getObjectFromJson(json: String?, type: Type): T? = GSON.fromJson<T?>(json, type)

}
